File: core/src/test/java/com/pipai/utils/CollisionUtilsTest.java
package com.pipai.utils;

import org.junit.Assert;
import org.junit.Test;

import com.badlogic.gdx.math.Vector2;
import com.pipai.adv.artemis.components.CollisionBounds;

public class CollisionUtilsTest {

    private static final CollisionBounds.CollisionBoundingBox BOUNDING_BOX = new CollisionBounds.CollisionBoundingBox(
            0, 0, 10, 10);

    @Test
    public void testOverlapRectRect() {
        Assert.assertTrue(CollisionUtils.INSTANCE.overlaps(
                0, 0, BOUNDING_BOX,
                9, 0, BOUNDING_BOX));

        Assert.assertTrue(CollisionUtils.INSTANCE.overlaps(
                9, 0, BOUNDING_BOX,
                0, 0, BOUNDING_BOX));

        Assert.assertTrue(CollisionUtils.INSTANCE.overlaps(
                0, 0, BOUNDING_BOX,
                0, 9, BOUNDING_BOX));

        Assert.assertTrue(CollisionUtils.INSTANCE.overlaps(
                0, 9, BOUNDING_BOX,
                0, 0, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                0, 0, BOUNDING_BOX,
                10, 0, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                10, 0, BOUNDING_BOX,
                0, 0, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                0, 0, BOUNDING_BOX,
                0, 10, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                0, 10, BOUNDING_BOX,
                0, 0, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                0, 0, BOUNDING_BOX,
                11, 0, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                11, 0, BOUNDING_BOX,
                0, 0, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                0, 0, BOUNDING_BOX,
                0, 11, BOUNDING_BOX));

        Assert.assertFalse(CollisionUtils.INSTANCE.overlaps(
                0, 11, BOUNDING_BOX,
                0, 0, BOUNDING_BOX));
    }

    @Test
    public void testMtvRectRect() {
        Assert.assertEquals(new Vector2(-1, 0),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        9, 0, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(1, 0),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        -9, 0, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(0, -1),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        0, 9, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(0, 1),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        0, -9, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(-1, 0),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        9, 8, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(1, 0),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        -9, -8, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(0, -1),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        8, 9, BOUNDING_BOX));

        Assert.assertEquals(new Vector2(0, 1),
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        -8, -9, BOUNDING_BOX));

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        10, 0, BOUNDING_BOX));

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        -10, 0, BOUNDING_BOX));

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        0, 10, BOUNDING_BOX));

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.INSTANCE.minimumTranslationVector(
                        0, 0, BOUNDING_BOX,
                        0, -10, BOUNDING_BOX));
    }

}

File: core/src/test/java/com/pipai/adv/utils/CollisionUtilsTest.kt
package com.pipai.adv.utils

import org.junit.Assert
import org.junit.Test

import com.badlogic.gdx.math.Vector2
import com.pipai.adv.artemis.components.CollisionBounds

class CollisionUtilsTest {

    @Test
    fun testOverlapRectRect() {
        Assert.assertTrue(CollisionUtils.overlaps(
                0f, 0f, BOUNDING_BOX,
                9f, 0f, BOUNDING_BOX))

        Assert.assertTrue(CollisionUtils.overlaps(
                9f, 0f, BOUNDING_BOX,
                0f, 0f, BOUNDING_BOX))

        Assert.assertTrue(CollisionUtils.overlaps(
                0f, 0f, BOUNDING_BOX,
                0f, 9f, BOUNDING_BOX))

        Assert.assertTrue(CollisionUtils.overlaps(
                0f, 9f, BOUNDING_BOX,
                0f, 0f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                0f, 0f, BOUNDING_BOX,
                10f, 0f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                10f, 0f, BOUNDING_BOX,
                0f, 0f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                0f, 0f, BOUNDING_BOX,
                0f, 10f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                0f, 10f, BOUNDING_BOX,
                0f, 0f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                0f, 0f, BOUNDING_BOX,
                11f, 0f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                11f, 0f, BOUNDING_BOX,
                0f, 0f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                0f, 0f, BOUNDING_BOX,
                0f, 11f, BOUNDING_BOX))

        Assert.assertFalse(CollisionUtils.overlaps(
                0f, 11f, BOUNDING_BOX,
                0f, 0f, BOUNDING_BOX))
    }

    @Test
    fun testMtvRectRect() {
        Assert.assertEquals(Vector2(-1f, 0f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        9f, 0f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(1f, 0f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        -9f, 0f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(0f, -1f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        0f, 9f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(0f, 1f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        0f, -9f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(-1f, 0f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        9f, 8f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(1f, 0f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        -9f, -8f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(0f, -1f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        8f, 9f, BOUNDING_BOX))

        Assert.assertEquals(Vector2(0f, 1f),
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        -8f, -9f, BOUNDING_BOX))

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        10f, 0f, BOUNDING_BOX))

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        -10f, 0f, BOUNDING_BOX))

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        0f, 10f, BOUNDING_BOX))

        Assert.assertEquals(Vector2.Zero,
                CollisionUtils.minimumTranslationVector(
                        0f, 0f, BOUNDING_BOX,
                        0f, -10f, BOUNDING_BOX))
    }

    companion object {

        private val BOUNDING_BOX = CollisionBounds.CollisionBoundingBox(
                0f, 0f, 10f, 10f)
    }

}
